datasets/covost2/covost2.py

```diff
@@ -0,0 +1,167 @@
+
+# fmt: off
+XX_EN_LANGUAGES = ["fr", "de", "es", "ca", "it", "ru", "zh-CN", "pt", "fa", "et", "mn", "nl", "tr", "ar", "sv-SE", "lv", "sl", "ta", "ja", "id", "cy"]
+EN_XX_LANGUAGES = ["de", "tr", "fa", "sv-SE", "mn", "zh-CN", "cy", "ca", "sl", "et", "id", "ar", "ta", "lv", "ja"]
+# fmt: on
+
+COVOST_URL_TEMPLATE = "https://dl.fbaipublicfiles.com/covost/covost_v2.{src_lang}_{tgt_lang}.tsv.tar.gz"
+
+
+def _get_builder_configs():
+    builder_configs = [
+        datasets.BuilderConfig(name=f"en_{lang}", version=datasets.Version(_VERSION)) for lang in EN_XX_LANGUAGES
+    ]
+
+    builder_configs += [
+        datasets.BuilderConfig(name=f"{lang}_en", version=datasets.Version(_VERSION)) for lang in XX_EN_LANGUAGES
+    ]
+    return builder_configs
+
+
+class Covost2(datasets.GeneratorBasedBuilder):
+    """CoVOST2 Dataset."""
+
+    VERSION = datasets.Version(_VERSION)
+
+    BUILDER_CONFIGS = _get_builder_configs()
+
+    @property
+    def manual_download_instructions(self):
+        return """\
+        You should download the Common Voice v4 dataset from https://commonvoice.mozilla.org/en/datasets.
+        and unpack it to a path `{COVOST_ROOT}/{SOURCE_LANG_ID}` and then pass the `{COVOST_ROOT}` path as `data_dir`
+        """
+
+    def _info(self):
+        return datasets.DatasetInfo(
+            description=_DESCRIPTION,
+            features=datasets.Features(
+                client_id=datasets.Value("string"),
+                file=datasets.Value("string"),
+                sentence=datasets.Value("string"),
+                translation=datasets.Value("string"),
+                id=datasets.Value("string"),
+            ),
+            supervised_keys=("file", "translation"),
+            homepage=_HOMEPAGE,
+            citation=_CITATION,
+        )
+
+    def _split_generators(self, dl_manager):
+        data_root = os.path.abspath(os.path.expanduser(dl_manager.manual_dir))
+        source_lang, target_lang = self.config.name.split("_")
+        source_path = os.path.join(data_root, source_lang)
+
+        if not os.path.exists(source_path):
+            raise FileNotFoundError("source path not found")
+
+        covost_url = COVOST_URL_TEMPLATE.format(src_lang=source_lang, tgt_lang=target_lang)
+        extracted_path = dl_manager.download_and_extract(covost_url)
+
+        covost_tsv_path = os.path.join(extracted_path, f"covost_v2.{source_lang}_{target_lang}.tsv")
+        cv_tsv_path = os.path.join(source_path, "validated.tsv")
+
+        covost_tsv = self._load_df_from_tsv(covost_tsv_path)
+        cv_tsv = self._load_df_from_tsv(cv_tsv_path)
+
+        df = pd.merge(
+            left=cv_tsv[["path", "sentence", "client_id"]],
+            right=covost_tsv[["path", "translation", "split"]],
+            how="inner",
+            on="path",
+        )
+
+        train_df = df[(df["split"] == "train") | (df["split"] == f"train_covost")]
+        valid_df = df[df["split"] == "dev"]
+        test_df = df[df["split"] == "test"]
```

Comment:
We try not to process dataframes in _split_generators but rather in _generate_examples.

_split_generators just makes sure that we have downloaded all the required resources, and _generate_examples load them and iterate on them to generate examples. If you don't mind can you move this part in _generate_examples ?